hal/stm32c0.c

```diff
@@ -0,0 +1,373 @@
+        FLASH_CR = reg | ((p >> FLASH_PAGE_SIZE_SHIFT) << FLASH_CR_PNB_SHIFT) | FLASH_CR_PER;
+        DMB();
+        FLASH_CR |= FLASH_CR_STRT;
+        flash_wait_complete();
+        FLASH_CR &= ~FLASH_CR_PER;
+    }
+    return 0;
+}
+
+static void clock_pll_off(void)
+{
+    uint32_t reg32;
+
+    /* Select HSISYS as SYSCLK source. */
+    reg32 = RCC_CFGR;
+    reg32 &= ~((1 << 1) | (1 << 0));
+    RCC_CFGR = (reg32 | RCC_CFGR_SW_HSISYS);
+    DMB();
+#if 0
+    /* Turn off PLL */
+    RCC_CR &= ~RCC_CR_PLLON;
+    DMB();
+#endif
+}
+
+/* This implementation will setup HSI RC 48 MHz as System Clock Source, set
+ * flash wait state to 2, and set all peripherals to 16MHz (div4)
+ *
+ * */
+static void clock_pll_on(int powersave)
+{
+    uint32_t reg32;
+    uint32_t cpu_freq, plln, pllm, pllq, pllp, pllr, hpre, ppre, flash_waitstates;
+
+    /* Enable Power controller */
+    APB1_CLOCK_ER |= PWR_APB1_CLOCK_ER_VAL;
+
+    /* Select clock parameters (CPU Speed = 48MHz) */
+    cpu_freq = 48000000;
+    pllm = 4;
+    plln = 80;
+    pllp = 10;
+    pllq = 5;
+    pllr = 5;
+    hpre  = RCC_PRESCALER_DIV_NONE;
+    ppre = RCC_PRESCALER_DIV_NONE;
+    flash_waitstates = 1;
+
+    flash_set_waitstates(flash_waitstates);
+
+    /* Enable internal high-speed oscillator. */
+    RCC_CR |= RCC_CR_HSION;
+    DMB();
+    while ((RCC_CR & RCC_CR_HSIRDY) == 0) {};
+
+    /* Select HSISYS as SYSCLK source. */
+    reg32 = RCC_CFGR;
+    reg32 &= ~((1 << 1) | (1 << 0));
+    RCC_CFGR = (reg32 | RCC_CFGR_SW_HSISYS);
+    DMB();
+
+    #if 0
+    /* Disable PLL */
+    RCC_CR &= ~RCC_CR_PLLON;
+
+    /* Set prescalers for AHB, ADC, ABP1, ABP2. */
+    reg32 = RCC_CFGR;
+    reg32 &= ~(0xF0); /* don't change bits [0-3] that were previously set */
+    RCC_CFGR = (reg32 | (hpre << 8)); /* RM0444 - 5.4.3 - RCC_CFGR */
+    DMB();
+    reg32 = RCC_CFGR;
+    reg32 &= ~(0x1C00); /* don't change bits [0-14] */
+    RCC_CFGR = (reg32 | (ppre << 12));  /* RM0444 - 5.4.3 - RCC_CFGR */
+    DMB();
+
+    /* Set PLL config */
+    reg32 = RCC_PLLCFGR;
+    reg32 |= RCC_PLLCFGR_PLLSRC_HSI16;
+    reg32 |= ((pllm - 1) << 4);
+    reg32 |= plln << 8;
+    reg32 |= ((pllp - 1) << 17);
+    reg32 |= ((pllr - 1) << 29);
+    RCC_PLLCFGR = reg32;
+
+    DMB();
+    /* Enable PLL oscillator and wait for it to stabilize. */
+    RCC_PLLCFGR |= RCC_PLLCFGR_PLLR_EN;
+    RCC_CR |= RCC_CR_PLLON;
+    DMB();
+    while ((RCC_CR & RCC_CR_PLLRDY) == 0) {};
+
+    /* Select PLL as SYSCLK source. */
+    reg32 = RCC_CFGR;
+    reg32 &= ~((1 << 1) | (1 << 0));
+    RCC_CFGR = (reg32 | RCC_CFGR_SW_PLL);
+    DMB();
+
+    /* Wait for PLL clock to be selected. */
+    while ((RCC_CFGR & ((1 << 1) | (1 << 0))) != RCC_CFGR_SW_PLL) {};
+#endif
+    /* SYSCFG, COMP and VREFBUF clock enable */
+    APB2_CLOCK_ER |= SYSCFG_APB2_CLOCK_ER_VAL;
+}
+
+void hal_init(void)
+{
+    clock_pll_on(0);
+}
+
+#ifdef FLASH_SECURABLE_MEMORY_SUPPORT
+static void RAMFUNCTION do_secure_boot(void)
```

Comment:
Please confirm `SEC_SIZE` is the same for C0 (I think it is). `SEC_SIZE` indicates the number of 2KB sectors from the start of flash to protect after jump.

Reply:
Correct. Checked RM0490.